### PR TITLE
Breakout Changes - Owen McIntosh

### PR DESCRIPTION
**Pull request by OwenMcintosh**
20 minutes to familiarise with pre-existing codebase
10 minutes to fix circular dependency issue
20 minutes to add mouse movement to paddle
1 hour to add restart via any key/mouse press after loss via gamestate data distinctions (various implementations tested, much prefer the final implementation)
40 minutes to change ball effect display from text to bar (had to familiarise myself with how I did this in the past)
15 minutes to create new "smallBall" powerup effect
2 hours to create audio manager system and find suitable assets
10 minutes to create screen shake functionality upon life-loss

Would've liked to have made the manager systems singletons, but I can neither remember how to do so, nor do I think it'd be appropriate to change such deep components of the overall project architecture.

### CODE REVIEW

File: Breakout/UI.h [hunk 1]
```diff
@@ -13,19 +13,39 @@ class UI
 	UI(sf::RenderWindow* window, int lives, GameManager* gameManager);
 	~UI();
 
-	void updatePowerupText(std::pair<POWERUPS, float>);
+	void update(float dt, std::pair<POWERUPS, float> powerup);
+
 	void lifeLost(int lives);
+	void HandleBarScale(float powerupTimeRemaining, sf::Color powerupColour);
```
Comment (RubenMateusPT) [inline Breakout/UI.h:19]: - To keep consistent with rest of the code, function should start lowercased
- powerupColour could be passed as a reference, since we dont modified inside the function

File: Breakout/UI.cpp [hunk 2]
```diff
@@ -18,63 +18,137 @@ UI::UI(sf::RenderWindow* window, int lives, GameManager* gameManager)
...
+	_powerupBarBackground.setPosition(875, 20);
+	_powerupBarBackground.setSize(sf::Vector2f(DEFAULT_BAR_X, DEFAULT_BAR_Y));
+	_powerupBarBackground.setFillColor(sf::Color(0, 0, 0, 0));
+
+	_powerupBar.setPosition(875, 20);
+	_powerupBar.setSize(sf::Vector2f(DEFAULT_BAR_X, DEFAULT_BAR_Y));
+	_powerupBar.setFillColor(sf::Color(0, 0, 0, 0));
+
+	_screenIsShaking = false;
+	_screenShakeTimer = SCREEN_SHAKE_DURATION;
```
Comment (RubenMateusPT) [inline Breakout/UI.cpp:35]: - Good work on adding constant variables to easily modify the game, as an extra step you could've done the same for setPosition

File: Breakout/UI.cpp [hunk 3]
```diff
@@ -18,63 +18,137 @@ UI::UI(sf::RenderWindow* window, int lives, GameManager* gameManager)
...
+		screenShake(dt);
+	}
+
+}
 
 void UI::updatePowerupText(std::pair<POWERUPS, float> powerup)
 {
 	std::ostringstream oss;
-
+	_powerupBarBackground.setFillColor(sf::Color(144, 144, 144, 255));
```
Comment (RubenMateusPT) [inline Breakout/UI.cpp:56]: - Could be a const variable to easily change the background color

File: Breakout/UI.cpp [hunk 4]
```diff
@@ -18,63 +18,137 @@ UI::UI(sf::RenderWindow* window, int lives, GameManager* gameManager)
...
+		_powerupText.setString("fire");
 		_powerupText.setFillColor(extraBallEffectsColour);
+		HandleBarScale(powerup.second, extraBallEffectsColour);
+		break;
+	case smallBall:
+		oss << std::fixed << std::setprecision(2) << powerup.second;
+		_powerupText.setString("scale");
+		_powerupText.setFillColor(smallBallEffectsColour);
+		HandleBarScale(powerup.second, smallBallEffectsColour);
 		break;
```
Comment (RubenMateusPT) [inline Breakout/UI.cpp:94]: - An alternative way to set the powerup text would be for the powerup base calss to contain a powerupName variable, that then each powerup would change to their name, this way we would we could do _powerupText.setString(powerUp.name) only once. Letting us not have to come to this switch statment to modify each powerup text change in case the powerups change their name.

File: Breakout/UI.h [hunk 5]
```diff
@@ -13,19 +13,39 @@ class UI
 	UI(sf::RenderWindow* window, int lives, GameManager* gameManager);
 	~UI();
 
-	void updatePowerupText(std::pair<POWERUPS, float>);
+	void update(float dt, std::pair<POWERUPS, float> powerup);
```
Comment (RubenMateusPT) [inline Breakout/UI.h:16]: - Thank you for adding an update function to the UI class, will make handling the UI better

File: Breakout/UI.h [hunk 6]
```diff
@@ -13,19 +13,39 @@ class UI
...
+	static constexpr int8_t SCREEN_SHAKE_OFFSET_X_MAX = 100;
+	static constexpr int8_t SCREEN_SHAKE_OFFSET_X_MIN = 50;
+
+	static constexpr int8_t SCREEN_SHAKE_OFFSET_Y_MAX = 100;
+	static constexpr int8_t SCREEN_SHAKE_OFFSET_Y_MIN = 50;
+
+	static constexpr float DEFAULT_BAR_X = 100.f;
+	static constexpr float DEFAULT_BAR_Y = 20.f;
 	static constexpr float LIFE_RADIUS = 15.0f;
 	static constexpr float LIFE_PADDING = 20.0f;
```
Comment (RubenMateusPT) [inline Breakout/UI.h:50]: - Lots of modafiable const variables. Godd work. (Could use comments to separate sections)

File: Breakout/PowerupSmallBall.h [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+{
+
+public:
+    PowerupSmallBall(sf::RenderWindow* window, Paddle* paddle, Ball* ball);
+    ~PowerupSmallBall();
+
+    std::pair<POWERUPS, float> applyEffect() override; // Method to apply the power-up effect
+
+private:
+    static int constexpr smallBallChance = 50;
```
Comment (RubenMateusPT) [inline Breakout/PowerupSmallBall.h:14]: - Variable name should be consistent: SMALL_BALL_CHANCE

File: Breakout/PowerupSmallBall.cpp [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+	_sprite.setFillColor(smallBallEffectsColour); // Same colour as SmallPaddle
+}
+
+PowerupSmallBall::~PowerupSmallBall()
+{
+}
+
+std::pair<POWERUPS, float> PowerupSmallBall::applyEffect()
+{
+	float timeLength = 5.f;
```
Comment (RubenMateusPT) [inline Breakout/PowerupSmallBall.cpp:15]: - The value "5.f" should be a const variable to easily change how long the affect should be applied

File: Breakout/PowerupBase.h [hunk 9]
```diff
@@ -6,8 +6,6 @@
 #include "Ball.h"
 #include <vector>
 
-#include "PowerupFireBall.h"
```
Comment (RubenMateusPT) [inline Breakout/PowerupBase.h:9]: - Good catch on removing circular dependency and unused include

File: Breakout/GameManager.h [hunk 10]
```diff
@@ -7,13 +7,23 @@
...
 #include "UI.h"
+#include "AudioManager.h"
 
+enum class GAMESTATE {
 
+    play = 0,
+    win = 1,
+    loss = 2
+
+};
```
Comment (RubenMateusPT) [inline Breakout/GameManager.h:18]: - Good enum to keep track of gamestate

File: Breakout/GameManager.cpp [hunk 11]
```diff
@@ -21,28 +21,61 @@ void GameManager::initialize()
...
     _ui = new UI(_window, _lives, this);
 
     // Create bricks
     _brickManager->createBricks(5, 10, 80.0f, 30.0f, 5.0f);
+    
+    // play music audio
+    _audioManager->playMusicByName("music");
+
+    // rng random seed
+    srand(time(0));
```
Comment (RubenMateusPT) [inline Breakout/GameManager.cpp:35]: - Good catch, this makes sure its truly random

File: Breakout/GameManager.cpp [hunk 12]
```diff
@@ -21,28 +21,61 @@ void GameManager::initialize()
...
+
+    delete _paddle;
+    delete _brickManager;
+    delete _messagingSystem;
+    delete _ball;
+    delete _powerupManager;
+    delete _audioManager;
+    delete _ui;
+
+    initialize();
```
Comment (RubenMateusPT) [inline Breakout/GameManager.cpp:57]: - This is fine since it only happens when we do a reset of the game. Altough It can still be a bit expensive, instead of deleting and then creating the objects again, we could add a reset function to each object that puts them on their initial state.

File: Breakout/GameManager.cpp [hunk 13]
```diff
@@ -21,28 +21,61 @@ void GameManager::initialize()
...
-
     if (_lives <= 0)
     {
-        _masterText.setString("Game over.");
+        _masterText.setString("Game over.\nPress Any Key To Restart.");
+        _gameState = GAMESTATE::loss;
         return;
     }
+
+    if (sf::Keyboard::isKeyPressed(sf::Keyboard::M)) _levelComplete = true;
```
Comment (RubenMateusPT) [inline Breakout/GameManager.cpp:73]: - Is this debug code to test game state? Need to be removed before going into production

File: Breakout/GameManager.cpp [hunk 14]
```diff
@@ -78,8 +111,8 @@ void GameManager::update(float dt)
     }
 
     // move paddle
-    if (sf::Keyboard::isKeyPressed(sf::Keyboard::D)) _paddle->moveRight(dt);
-    if (sf::Keyboard::isKeyPressed(sf::Keyboard::A)) _paddle->moveLeft(dt);
+    if (sf::Keyboard::isKeyPressed(sf::Keyboard::D) || sf::Mouse::isButtonPressed(sf::Mouse::Right)) _paddle->moveRight(dt);
+    if (sf::Keyboard::isKeyPressed(sf::Keyboard::A) || sf::Mouse::isButtonPressed(sf::Mouse::Left)) _paddle->moveLeft(dt);
```
Comment (RubenMateusPT) [inline Breakout/GameManager.cpp:115]: - This is good, lets us have more keys to use to control the paddle

File: Breakout/Ball.cpp [hunk 15]
```diff
@@ -77,17 +82,31 @@ void Ball::update(float dt)
...
+        return; // no collisisons when in fireBall mode.
+    }
     if (collisionResponse == 1)
     {
+        _gameManager->getAudioManager()->playSoundByName("hit");
         _direction.x *= -1; // Bounce horizontally
     }
     else if (collisionResponse == 2)
     {
+        _gameManager->getAudioManager()->playSoundByName("hit");
```
Comment (RubenMateusPT) [inline Breakout/Ball.cpp:109]: - Instead of having to write the name of the sound manually, they should be a const variable, this avoids someone misspeling the string and then wonder why the sound isnt playing

File: Breakout/GameManager.cpp [hunk 16]
```diff
@@ -21,28 +21,61 @@ void GameManager::initialize()
...
-    _powerupManager = new PowerupManager(_window, _paddle, _ball);
+    _audioManager = new AudioManager();
+    _powerupManager = new PowerupManager(_window, _paddle, _ball, _audioManager);
     _ui = new UI(_window, _lives, this);
 
     // Create bricks
     _brickManager->createBricks(5, 10, 80.0f, 30.0f, 5.0f);
+    
+    // play music audio
+    _audioManager->playMusicByName("music");
```
Comment (RubenMateusPT) [inline Breakout/GameManager.cpp:32]: - Instead of having to write the name of the sound manually, they should be a const variable, this avoids someone misspeling the string and then wonder why the sound isnt playing

File: Breakout/AudioManager.h [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+{
+    sf::SoundBuffer buffer;
+    std::string name;
+};
+
+struct musicFile 
+{
+    std::string filename;
+    std::string name;
+};
```
Comment (RubenMateusPT) [inline Breakout/AudioManager.h:14]: - Good thinking on creating struct to easily hold information on sound and music files

File: Breakout/AudioManager.cpp [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+	for (int i = 0; i < _soundVector.size(); i++) 
+	{
+		delete _soundVector[i];
+	}
+
+	for (int i = 0; i < _musicVector.size(); i++)
+	{
+		delete _musicVector[i];
+	}
+}
```
Comment (RubenMateusPT) [inline Breakout/AudioManager.cpp:24]: Good work making sure you cleanup any created objects

File: Breakout/AudioManager.cpp [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+	
+	soundFile* newSound = new soundFile();
+
+	if (!newSound->buffer.loadFromFile(fileName))
+	{
+		// Error...
+		printf("error loading sound");
+		delete newSound;
+		return;
+	}
```
Comment (RubenMateusPT) [inline Breakout/AudioManager.cpp:38]: - Good work adding error handling

File: Breakout/AudioManager.cpp [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+
+// given a music file name, and it's respective use-case name, load and store a new musicFile struct object
+void AudioManager::addMusic(std::string fileName, std::string musicName)
+{
+	sf::Music music;
+	if (!music.openFromFile(fileName))
+	{
+		// Error...
+		return;
+	}
```
Comment (RubenMateusPT) [inline Breakout/AudioManager.cpp:72]: - Good work adding error handling, but dont forget to add a print message to console/log ;)

File: Breakout/AudioManager.cpp (hunk 19 above)
Comment (RubenMateusPT) [reply Breakout/AudioManager.cpp:38]: - And delleting the object so it doesnt stay around!

File: Breakout/AudioManager.cpp (hunk 20 above)
Comment (RubenMateusPT) [reply Breakout/AudioManager.cpp:72]: - Dotn forget to delete the music object if you werent able to add the file! (Do what you did with the sound one)

File: Breakout/AudioManager.cpp [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+
+	for (int i = 0; i < _soundVector.size(); i++) 
+	{
+
+		if (_soundVector[i]->name == soundName)
+		{
+
+			_soundToBePlayed.setBuffer(_soundVector[i]->buffer);
+			_soundToBePlayed.play();
+
```
Comment (RubenMateusPT) [inline Breakout/AudioManager.cpp:57]: - Could add a break statement so the loop doesnt run after finding the sound needed to play

File: Breakout/AudioManager.cpp [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+	{
+
+		if (_musicVector[i]->name == musicName)
+		{
+
+			_currentMusic.openFromFile(_musicVector[i]->filename);
+			_currentMusic.setVolume(100);
+			_currentMusic.play();
+			_currentMusic.setLoop(true);
+
```
Comment (RubenMateusPT) [inline Breakout/AudioManager.cpp:93]: - Could add a break statement so the loop doesnt run after finding the music needed to play

File: Breakout/AudioManager.cpp [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+void AudioManager::FadeMusicOut(float dt) {
+
+	if (_currentMusic.getVolume() - FADE_OUT_INTENSITY * dt <= 0.f) {
+		_currentMusic.setVolume(0);
+		return;
+	}
+
+	_currentMusic.setVolume(_currentMusic.getVolume() - FADE_OUT_INTENSITY * dt);
+}
```
Comment (RubenMateusPT) [inline Breakout/AudioManager.cpp:110]: - Nice Fadeout Music logic

File: Breakout/UI.cpp [hunk 24]
```diff
@@ -18,63 +18,137 @@ UI::UI(sf::RenderWindow* window, int lives, GameManager* gameManager)
...
 	_lives[_lives.size() - 1 - lives].setFillColor(sf::Color::Transparent);
+	_screenIsShaking = true;
+}
+
+void UI::HandleBarScale(float powerupTimeRemaining, sf::Color powerupColour) {
+	_powerupBar.setSize(sf::Vector2f((powerupTimeRemaining / 5.f) * _powerupBarBackground.getSize().x, _powerupBarBackground.getSize().y));
+	_powerupBar.setFillColor(powerupColour);
+}
+
+void UI::screenShake(float dt)
```
Comment (RubenMateusPT) [inline Breakout/UI.cpp:116]: Good ScreenShake

**Review by RubenMateusPT — COMMENTED** [review]
Good Audio Manager and Screen Shake Implementation.
Some kinks in one or another place, but overall good code structure and nice readability.
